src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java

```diff
@@ -1324,6 +1244,145 @@ public boolean isKnownParticipant( UUID organizationId, UUID studyId, String par
+
+                if ( studiesESID == null || participantsESID == null )
+                    return;
+
+                Iterable<SetMultimap<FullQualifiedName, Object>> studyEntities = dataApi
+                        .loadSelectedEntitySetData(
+                                studiesESID,
+                                new EntitySetSelection(
+                                        Optional.of( ImmutableSet.of(
+                                                propertyTypeIdsByFQN.get( STRING_ID_FQN )
+                                        ) )
+                                ),
+                                FileType.json
+                        );
+
+                // map studyIds -> studyEKIDs
+                StreamUtil.stream( studyEntities ).map( Multimaps::asMap ).forEach( ( entity ) -> {
+                    UUID studyEKID = getFirstUUIDOrNull( entity, ID_FQN );
+                    UUID studyId = getFirstUUIDOrNull( entity, STRING_ID_FQN );
+                    if ( studyId == null )
+                        return;
+
+                    Map<UUID, UUID> mappedEntities = studyEntityKeyIds.getOrDefault( orgId, Maps.newHashMap() );
+                    mappedEntities.put( studyId, studyEKID );
+                    studyEntityKeyIds.put( orgId, mappedEntities );
+                    studyIds.put( studyEKID, studyId );
+                } );
+
+                // get study neighbors constrained by used_by and participated_in associations
+                Set<UUID> edgeESIDS = Sets.newHashSet( participatedInESID, usedByESID )
+                        .stream().filter( Objects::nonNull )
+                        .collect( Collectors.toSet() );
+
+                Set<UUID> srcESIDS = Sets.newHashSet( deviceESID, participantsESID )
+                        .stream().filter( Objects::nonNull )
+                        .collect( Collectors.toSet() );
+
+                Map<UUID, List<NeighborEntityDetails>> studyNeighbors = searchApi.executeFilteredEntityNeighborSearch(
+                        studiesESID,
+                        new EntityNeighborsFilter(
+                                studyIds.keySet(),
+                                Optional.of( srcESIDS ),
+                                Optional.empty(),
+                                Optional.of( edgeESIDS )
+                        )
+                );
+
+                // process neighbors
+                Map<UUID, Map<String, UUID>> participantsByStudy = Maps
+                        .newHashMap(); // studyId -> participantId -> EKID
+                studyNeighbors.forEach( ( studyEKID, neighbors ) -> {
+                    UUID studyId = studyIds.get( studyEKID );
+
+                    neighbors.forEach( neighbor -> {
+                        if ( neighbor.getAssociationEntitySet().getId().equals( usedByESID ) ) {
+                            String deviceId = getFirstValueOrNull( neighbor.getNeighborDetails().get(), STRING_ID_FQN );
```

Comment:
`neighbor.getNeighborDetails()` is optional, so it's possible for `.get()` to actually throw if the value is null.